src/ESPeasyGPIO.ino

```diff
@@ -0,0 +1,393 @@
+/****************************************************************************/
+// Central functions for GPIO handling
+// **************************************************************************/
+#include "EspEasyGPIO.h"
+
+//********************************************************************************
+// Internal GPIO write
+//********************************************************************************
+void GPIO_Internal_Write(byte pin, byte value)
+{
+  pinMode(pin, OUTPUT);
+  digitalWrite(pin, value);
+}
+
+//********************************************************************************
+// Internal GPIO read
+//********************************************************************************
+bool GPIO_Internal_Read(byte pin)
+{
+  return digitalRead(pin)==HIGH;
+}
+
+boolean GPIO_Read_Switch_State(struct EventStruct *event) {
+  const byte pinNumber     = CONFIG_PIN1;
+  const uint32_t key = createKey(PLUGIN_GPIO, pinNumber);
+
+  if (existPortStatus(key)) {
+    return GPIO_Read_Switch_State(pinNumber, globalMapPortStatus[key].mode);
+  }
+  return false;
+}
+
+boolean GPIO_Read_Switch_State(byte pinNumber, byte pinMode) {
+  bool canRead = false;
+
+  switch (pinMode)
+  {
+    case PIN_MODE_UNDEFINED:
+    case PIN_MODE_INPUT:
+    case PIN_MODE_INPUT_PULLUP:
+    case PIN_MODE_OUTPUT:
+      canRead = true;
+      break;
+    case PIN_MODE_PWM:
+      break;
+    case PIN_MODE_SERVO:
+      break;
+    case PIN_MODE_OFFLINE:
+      break;
+    default:
+      break;
+  }
+
+  if (!canRead) { return false; }
+
+  // Do not read from the pin while mode is set to PWM or servo.
+  // See https://github.com/letscontrolit/ESPEasy/issues/2117#issuecomment-443516794
+  return digitalRead(pinNumber) == HIGH;
+}
+
+//********************************************************************************
+// MCP23017 read
+//********************************************************************************
+int8_t GPIO_MCP_Read(byte Par1)
+{
+  int8_t state = -1;
+  byte unit = (Par1 - 1) / 16;
+  byte port = Par1 - (unit * 16);
+  uint8_t address = 0x20 + unit;
+  byte IOBankValueReg = 0x12;
+  if (port > 8)
+  {
+    port = port - 8;
+    IOBankValueReg++;
+  }
+  // get the current pin status
+  Wire.beginTransmission(address);
+  Wire.write(IOBankValueReg); // IO data register
+  Wire.endTransmission();
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    state = ((Wire.read() & _BV(port - 1)) >> (port - 1));
+  }
+  return state;
+}
+
+
+//********************************************************************************
+// MCP23017 write
+//********************************************************************************
+boolean GPIO_MCP_Write(byte Par1, byte Par2)
+{
+  boolean success = false;
+  byte portvalue = 0;
+  byte unit = (Par1 - 1) / 16;
+  byte port = Par1 - (unit * 16);
```

Comment:
No check for ranges in port and unit?
And this looks like a div/mod split, right?

Reply:
It's not my code. It's the original code.
I just moved it outside the plugin.
Checks are performed in the single commands, in order to throw log errors.